Firestore/src/FirestoreSessionHandler.php

```diff
@@ -0,0 +1,404 @@
+     * @var string
+     */
+    private $savePath;
+    /**
+     * @var string
+     */
+    private $sessionName;
+    /**
+     * @var Transaction
+     */
+    private $transaction;
+
+    /**
+     * Create a custom session handler backed by Cloud Firestore.
+     *
+     * @param ConnectionInterface $connection A Connection to Cloud Firestore.
+     * @param ValueMapper $valueMapper A Firestore Value Mapper.
+     * @param string $projectId The current project id.
+     * @param string $database The database id.
+     * @param array $options [optional]
+     */
+    public function __construct(
+        ConnectionInterface $connection,
+        ValueMapper $valueMapper,
+        $projectId,
+        $database,
+        array $options = []
+    ) {
+        $this->connection = $connection;
+        $this->valueMapper = $valueMapper;
+        $this->projectId = $projectId;
+        $this->database = $database;
+        $this->options = $options + [
+            'begin' => [],
+            'commit' => [],
+            'rollback' => [],
+            'delete' => [],
+            'query' => [],
+            'gcLimit' => 0,
+            'collectionNameTemplate' => '%1$s:%2$s',
+        ];
+
+        // Cut down gcLimit to 1000
+        $this->options['gcLimit'] = min($this->options['gcLimit'], 1000);
+    }
+
+    /**
+     * Start a session, by getting the Firebase collection from $savePath.
+     *
+     * @param string $savePath The value of `session.save_path` setting will be
+     *        used here as the Firestore collection ID.
+     * @param string $sessionName The value of `session.name` setting will be
+     *        used here as the Firestore document ID prefix (ex: "PHPSESSID:").
+     * @return bool
+     */
+    public function open($savePath, $sessionName)
+    {
+        $this->savePath = $savePath;
+        $this->sessionName = $sessionName;
+        $database = $this->databaseName($this->projectId, $this->database);
+
+        $beginTransaction = $this->connection->beginTransaction([
+            'database' => $database
+        ] + $this->options['begin']);
+
+        $this->transaction = new Transaction(
+            $this->connection,
+            $this->valueMapper,
+            $database,
+            $beginTransaction['transaction']
+        );
+
+        return true;
+    }
+
+    /**
+     * Just return true for this implementation.
+     */
+    public function close()
+    {
+        return true;
+    }
+
+    /**
+     * Read the session data from Cloud Firestore.
+     *
+     * @param string $id Identifier used for the session.
+     * @return string
+     */
+    public function read($id)
+    {
+        try {
+            $docRef = $this->getDocumentReference(
+                $this->connection,
+                $this->valueMapper,
+                $this->projectId,
+                $this->database,
+                $this->docId($id)
+            );
+            $snapshot = $this->transaction->snapshot($docRef);
```

Comment:
user-specified options from `__construct()`'s `$options` would be good for this method.